docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+
+  tokens.foreach(access) // Emit `tokens.size` copies of `access(...)`, putting a token on each molecule.
+  val result: () => T = () => request()
+  result
+}
+// Example usage:
+val doWorkWithTwoAccesses = wrapWithAccessTokens(Set("token1", "token2"), t => println(s"do work with token $t"))
+// Now `doWork()` will be called by at most 2 processes at a time.
+
+// ... start a new process, in which:
+doWorkWithTwoAccesses()
+
+```
+
+When concurrent processes emit `request()` molecules, only at most `n` processes will actually do work on the resource at any one time.
+Each process will be assigned a token out of the available set of tokens.
+
+## Concurrent critical sections, or `Object.synchronized`
+
+A "critical section" is a portion of code that cannot be safely called from different processes at the same time.
+We will now implement the following requirements:
+
+- The calls `beginCritical()` and `endCritical()` can be made in any reaction, in this order.
+- The code between these two calls can be executed only by one reaction at a time, among all reactions that call these functions.
+- A reaction that calls `beginCritical()` will be blocked if another reaction already called `beginCritical()` but did not yet call `endCritical()`, and will be unblocked only when that other reaction calls `endCritical`. 
+
+This functionality is similar to `Object.synchronized`, which provides synchronized access to a block of reentrant code.
+In our case, the critical section is delimited by two function calls `beginCritical()` and `endCritical()` are separate function calls that can be made at any two points in the code -- including `if` expressions or inside closures -- which is impossible with `synchronized` blocks.
+Also, the `Object.synchronized` construction identifies the synchronized blocks by an `Object` reference: different objects will be responsible for synchronizing different blocks of reentrant code.
+What we would like to allow is _any_ code anywhere to contain any number of critical sections identified by the same `beginCritical()` call.
+
+How can we implement this functionality using the chemical machine?
+Since the only way to communicate between reactions is to emit molecules, `beginCritical()` and `endCritical()` must be molecules.
+Clearly, `beginCritical` must be blocking while `endCritical` does not have to be; so let us make `endCritical` a non-blocking molecule.
+
+What should happen when a process emits `beginCritical()`?
+We must enforce a single-process access to the critical section.
+In other words, a reaction that consumes `beginCritical()` should only start one at a time even if several `beginCritical` molecules are available.
+Therefore, this reaction must also require another input molecule, say `access()`, of which we will only have a single copy emitted into the soup:
+
+```scala
+val beginCritical = b[Unit, Unit]
+val access = m[Unit]
+
+site(
+  go { case beginCritical(_, reply) + access(_) => ???; reply(); ??? }
```

Comment:
Cannot go be a simple function of two parameters, i.e. the input molecules that form the case as one parameter and the RHS or reaction body (emitting part), being the second parameter? Or does this reply business on blocking molecules make that impossible?